samples: basic: add multitarget_hello_world

gmarull opened the pull request:
Add a new "hello world" sample that can be built for multiple board
targets. This sample can be used to see if Zephyr boots successfully
with _minimal_ configuration on SoCs with multiple cores, or boards with
multiple SoCs.

The following configurations are provided for now:

- nRF5340 DK: cpuapp+cpunet
- nRF54H20 DK: cpuapp+cpurad, cpuapp+cpuppr, cpuapp+cpuppr(xip)

Note that because of a quoting bug somewhere in the build system, SB_CONFIG_REMOTE_BOARD is specified in .conf files (see the sysbuild/ directory). This should be fixed in the near future.

gmarull committed:
samples: sysbuild: add hello world for multiple board targets

Add a new "hello world" sample that can be built for multiple board
targets. This sample can be used to see if Zephyr boots successfully
with _minimal_ configuration on SoCs with multiple cores, or boards with
multiple SoCs.

The following configurations are provided for now:

- nRF5340 DK: cpuapp+cpunet
- nRF54H20 DK: cpuapp+cpurad, cpuapp+cpuppr, cpuapp+cpuppr(xip)

Note that because of a quoting bug somewhere in the build system,
SB_CONFIG_REMOTE_BOARD is specified in .conf files (see the sysbuild/
directory). This should be fixed in the near future.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

nordicjm commented on samples/basic/multicore_hello_world/sample.yaml:
```diff
@@ -0,0 +1,45 @@
+    platform_allow:
+      - nrf54h20pdk_nrf54h20_cpuapp
+    integration_platforms:
+      - nrf54h20pdk_nrf54h20_cpuapp
```
Needs porting to hwmv2

gmarull replied:
yup, will rebase after hwmv2 is merged

57300 commented on samples/basic/multitarget_hello_world/sample.yaml:
```diff
@@ -0,0 +1,45 @@
+sample:
+  name: Multicore/SoC Hello World
```
Rename?
```suggestion
  name: Multitarget Hello World
```

57300 commented on samples/basic/multitarget_hello_world/sample.yaml:
```diff
@@ -0,0 +1,45 @@
+  name: Multicore/SoC Hello World
+  description: |
+    Hello World application that builds for a remote core or SoC from sources.
+    Both images print board name they were run on.
```
```suggestion
    Both images print the board target they were run on.
```

57300 commented on samples/basic/multitarget_hello_world/sysbuild.cmake:
```diff
@@ -0,0 +1,15 @@
+  BOARD ${SB_CONFIG_REMOTE_BOARD}
+)
+
+add_dependencies(multitarget_hello_world remote)
+sysbuild_add_dependencies(FLASH multitarget_hello_world remote)
```
Is it important for `remote` to be built and flashed first?

57300 reviewed (commented):
LGTM, although I wonder if it would be better to add a multitarget option to regular `hello_world` instead.

gmarull commented:
@nordicjm ping

fabiobaltieri commented:
It's very tempting to suggest that this should have a multitarget blinky mode as well

kartben commented on samples/basic/multitarget_hello_world/README.rst:
```diff
@@ -0,0 +1,69 @@
+.. zephyr:code-sample:: multitarget_hello_world
+   :name: Hello World for multiple board targets
+
+   Run a hello world sample on multiple board targets
```
I am a bit torn as to whether that should be a basic sample, and I don't think advertising it using "Hello World" terminology is effective (think: what would people use as keywords to find this)

I would suggest to move this to application_development and, while of course the code itself can be a hello world, I think the name & description should be more descriptive of the use case (similar to a comment also made by @57300).

Something like the below:

```suggestion
   :name: Multi-target application

   Use sysbuild to build and run on multiple board targets
```

gmarull replied:
I'm fine moving the sample. Shouldn't we have `samples/sysbuild`? Then we could have it under `samples/sysbuild/hello_world`, and call it "Multi-target hello world application".

kartben replied:
We already have `samples/application_development/sysbuild/with_mcuboot` so maybe `samples/application_development/sysbuild/hello_world` and add an index + ToC in `samples/application_development/sysbuild/` folder so that it's more visible? Or also fine to move sysbuild as a top-level folder but my point is the new sample should probably live next to the existing sysbuild mcuboot one.
Thanks!

carlescufi replied:
No objections from me, both options make sense.

gmarull replied:
`samples/application_development/sysbuild/with_mcuboot` is an extremely long path, I'd suggest shortening things. `application_development` is also very generic...

something like
- `samples/sysbuild/mcuboot`
- `samples/sysbuild/hello_world`

feels better IMHO.

fabiobaltieri replied:
ok since you are rescoping this, how about making it blink led-0 and led-1 (at different rates) as well? just to give some confidence that the cores are alive and running for boards where the integrated debugger only has a single uart available

gmarull replied:
> ok since you are rescoping this, how about making it blink led-0 and led-1 (at different rates) as well? just to give some confidence that the cores are alive and running for boards where the integrated debugger only has a single uart available

sounds like a good sample, but shouldn't it be ab extra `samples/sysbuild/blinky`, following our basic samples?

gmarull replied:
I'll propose a samples/sysbuild/blinky if this moves on, not before, though.

gmarull committed:
samples: move with_mcuboot sample to samples/sysbuild

This creates a new category of samples that demonstrate certain features
of sysbuild. "application development" was a very generic name.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
last push: created a new category `samples/sysbuild` for samples that its main purpose is to demonstrate sysbuild features.

carlescufi commented:
@kartben please take another look

kartben commented on samples/sysbuild/with_mcuboot/README.rst:
I'm started to think we should find a way to automate this somehow as I feel like a PITA always asking for this but: could you please make sure this has a redirect?
+1 otherwise, very cool stuff, thanks!

gmarull replied:
done
